proposals/java_agents_and_more_layers.md

```diff
@@ -0,0 +1,280 @@
+
+## Motivation
+
+There are 3 feature requests that may be able to be solved together. These are:
+
+- Give users ability to add Java agents ([#378](/../../issues/378))
+- Give users ability to copy arbitrary files to the image ([#213](/../../issues/213))
+- Separate frequently changing from non-changing dependencies for more incrementality ([#403](/../../issues/403))
+
+## Synopsis of the issues
+
+### Give users ability to add Java agents ([#378](/../../issues/378))
+
+Currently, Jib only adds application files to the container image. These include dependency JARs, resource files, and classes. However, Java developers often run their applications with Java Agents that execute as part of the JVM invocation.
+
+For example, in [this Dockerfile example](https://github.com/saturnism/spring-petclinic-gcp/blob/master/docker/Dockerfile), the Cloud Debugger and Cloud Profiler agents are added to the container image. The agent archives are first downloaded, and then extracted to their own directory on the image. The archives contains the `.so` file to pass to the java invocation in the form of:
+
+```shell
+java -agentpath:/path/to/agent/files/someagent.so ...
+```
+
+### Give users ability to copy arbitrary files to the image ([#213](/../../issues/213))
+
+Users may wish to add other files for use in the image. Currently, the way to do this would be to place the files within the application resources. However, this conflates the classpath of the application, the file can only be reached under the `/app/resources` path, and changes to the file would mean repushing all the resources. Therefore, users should have some way of adding files to a new custom layer.
+
+### Separate frequently changing from non-changing dependencies for more incrementality ([#403](/../../issues/403))
+
+The dependencies layer tends to be the largest layer in the image, since it contains all the dependency JARs. However, not all dependencies are the same. Some may change more frequently than others (especially `-SNAPSHOT` versions). Therefore, users should have some way to group different dependencies into different layers. For example, separating released artifacts from snapshot artifacts, and grouping dependencies from a BOM into a layer, or separating more frequently changed ones from less frequently changed ones to reduce the amortized push time.
+
+## Proposal
+
+The proposal is to keep the configuration for adding additional files and Java agents separate from the configuration for separating the application layers into thinner layers. The point of this is to keep the configuration simple and not allow arbitrary user-controlled layering.
+
+The configuration for adding additional files (including Java agents) would look something like:
+
+*(The exact configuration is to be decided before this PR is merged. Alternative naming for this parameter include: `additionalFiles`, `extraFiles`, and `copyFiles`.)*
```

Comment:
Actually, I was just thinking and maybe we should have it be the semantics of [Dockerfile `COPY`](https://docs.docker.com/engine/reference/builder/#copy), where essentially:

- if `from` is a directory, only the contents are copied, and not the directory itself
- if `to` ends with a trailing slash, it is considered a directory. Otherwise, it is considered a regular file.
- if `from` refers to multiple files (like the contents of a directory), `to` must end with a trailing slash, and the `from` files will be added to that directory.

We won't support glob matching though, at least not until someone says they need it.

Reply:
Using Docker's COPY semantics makes sense.